File: frontend/src/api/endpoints/kpis.ts
/**
 * KPI API functions.
 *
 * Backend source: app/routers/kpis.py  (prefix: /kpis)
 * Endpoints consumed:
 *   GET /kpis/availability    → KpiResponse
 *   GET /kpis/utilization     → KpiResponse
 *   GET /kpis/critical-tasks  → CriticalTasksKpiResponse
 *
 * Shared query parameters (_kpi_query dependency):
 *   plan_id, corridor_id, department, horizon_type
 *
 * /kpis/critical-tasks additionally accepts:
 *   limit, offset  (pagination, default 50 / 0)
 *
 * IMPORTANT: KPI values (availability, utilization, critical task counts)
 * are computed exclusively by the backend. The frontend MUST NOT calculate,
 * derive, or synthesise these values. Always display what the backend returns.
 *
 * Rules:
 *   - All functions go through the shared apiClient.
 *   - No mock data, no fallbacks, no invented response fields.
 */

import { apiClient } from '../client';
import type {
  CriticalTasksKpiResponse,
  CriticalTasksParams,
  KpiQueryParams,
  KpiResponse,
} from '../types';

// ---------------------------------------------------------------------------
// GET /kpis/availability
// ---------------------------------------------------------------------------

/**
 * Asset availability KPI for a plan or horizon.
 *
 * Prototype formula (computed backend-side): scheduled_tasks / total_tasks * 100
 *
 * @param params – optional plan/corridor/department/horizon filter
 * @returns KpiResponse containing asset_availability_percent and related counts
 */
export async function getAvailabilityKpi(
  params?: KpiQueryParams,
): Promise<KpiResponse> {
  const response = await apiClient.get<KpiResponse>('/kpis/availability', {
    params,
  });
  return response.data;
}

// ---------------------------------------------------------------------------
// GET /kpis/utilization
// ---------------------------------------------------------------------------

/**
 * Window utilization KPI for a plan or horizon.
 *
 * Prototype formula (computed backend-side): scheduled_hours / available_window_hours * 100
 *
 * @param params – optional plan/corridor/department/horizon filter
 * @returns KpiResponse containing block_utilization_percent and hour totals
 */
export async function getUtilizationKpi(
  params?: KpiQueryParams,
): Promise<KpiResponse> {
  const response = await apiClient.get<KpiResponse>('/kpis/utilization', {
    params,
  });
  return response.data;
}

// ---------------------------------------------------------------------------
// GET /kpis/critical-tasks
// ---------------------------------------------------------------------------

/**
 * Paginated list of unscheduled severity-A tasks for a plan.
 *
 * @param params – optional KPI filters plus pagination (limit/offset)
 * @returns CriticalTasksKpiResponse (items array + PageMeta)
 */
export async function getCriticalTasksKpi(
  params?: CriticalTasksParams,
): Promise<CriticalTasksKpiResponse> {
  const response = await apiClient.get<CriticalTasksKpiResponse>(
    '/kpis/critical-tasks',
    { params },
  );
  return response.data;
}

File: backend/app/schemas/kpis.py
"""KPI query and response schemas."""

from __future__ import annotations

from datetime import date

from pydantic import Field

from app.schemas.common import ApiModel, CorridorId, EntityId, PageMeta
from app.schemas.enums import Department, PlanningHorizonType
from app.schemas.tasks import UnscheduledCriticalTaskResponse



class KpiQuery(ApiModel):
    plan_id: EntityId | None = None
    corridor_id: CorridorId | None = None
    department: Department | None = None
    horizon_type: PlanningHorizonType | None = None
    horizon_start: date | None = None



class KpiResponse(ApiModel):
    """Plan quality metrics. Prototype asset availability is scheduled/total * 100."""

    plan_id: EntityId | None = None
    horizon_type: PlanningHorizonType | None = None
    horizon_start: date | None = None
    horizon_end: date | None = None
    corridor_id: CorridorId | None = None
    department: Department | None = None
    total_tasks: int = Field(ge=0)
    scheduled_tasks: int = Field(ge=0)
    unscheduled_tasks: int = Field(ge=0)
    critical_unscheduled_tasks: int = Field(
        ge=0,
        description="Unscheduled tasks with defect severity A.",
    )
    asset_availability_percent: float = Field(
        ge=0,
        le=100,
        description=(
            "Prototype definition: scheduled_tasks / total_tasks * 100. "
            "Evaluates overall task fulfillment."
        ),
    )
    scheduled_hours: float = Field(default=0, ge=0)
    available_window_hours: float = Field(default=0, ge=0)
    block_utilization_percent: float = Field(
        default=0.0,
        ge=0,
        description="Window utilization definition: (scheduled_hours / available_window_hours) * 100.",
    )
    plan_generation_seconds: float | None = Field(
        default=None,
        ge=0,
        description="Duration in seconds taken by the ML + OR-Tools solver optimization pipeline.",
    )



class CriticalTasksKpiResponse(ApiModel):
    """Unscheduled critical work for the KPI surface."""

    items: list[UnscheduledCriticalTaskResponse]
    meta: PageMeta
